File: src/main/java/net/hectus/neobb/structure/Structure.java
package net.hectus.neobb.structure;

import com.google.gson.GsonBuilder;
import com.marcpg.libpg.storing.Cord;
import com.marcpg.libpg.storing.Pair;
import net.hectus.neobb.NeoBB;
import net.hectus.neobb.game.util.Arena;
import net.hectus.neobb.util.Utilities;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.World;
import org.bukkit.block.Block;
import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.Serializable;
import java.nio.file.Files;
import java.util.*;
import java.util.stream.Collectors;

public class Structure implements Serializable {
    public final String name;
    public final Set<Material> materials = new HashSet<>();
    public final BlockInfo[][][] blocks;
    public final int blocksX;
    public final int blocksY;
    public final int blocksZ;
    public transient Structure rotated;

    public Structure(String name, BlockInfo @NotNull [][][] blocks) {
        this.name = name;
        this.blocks = blocks;
        this.blocksX = blocks.length;
        this.blocksY = blocks[0].length;
        this.blocksZ = blocks[0][0].length;

        materials.addAll(Arrays.stream(blocks).parallel() // I'm actually not sure if bukkit supports parallel streams, but it shouldn't cause any issues even if it doesn't.
                .flatMap(Arrays::stream)
                .flatMap(Arrays::stream)
                .filter(Objects::nonNull)
                .map(BlockInfo::material)
                .collect(Collectors.toSet()));
    }

    public Structure(String name, World world, @NotNull Cord corner1, @NotNull Cord corner2) {
        this(name, blocks(world, corner1, corner2));
    }

    public void save() {
        String mode = Objects.requireNonNullElse(NeoBB.CONFIG.getString("structure-mode"), "local");
        if (mode.equals("server")) {
            NeoBB.LOG.warn("Saving structures to the server is not supported yet!");
        } else {
            File file = NeoBB.STRUCTURE_DIR.resolve(name + ".json").toFile();
            try (FileWriter writer = new FileWriter(file)) {
                new GsonBuilder().setPrettyPrinting().create().toJson(this, writer);
                NeoBB.LOG.info("Saved structure {} to {}", name, file.getAbsolutePath());
            } catch (IOException e) {
                NeoBB.LOG.error("Failed to save structure {} to {}", name, file.getAbsolutePath(), e);
            }
        }
    }

    public void remove() throws IOException {
        String mode = Objects.requireNonNullElse(NeoBB.CONFIG.getString("structure-mode"), "local");
        if (mode.equals("server")) {
            NeoBB.LOG.warn("Removing structures from the server is not supported yet!");
        } else {
            Files.delete(NeoBB.STRUCTURE_DIR.resolve(name + ".json"));
            NeoBB.LOG.info("Removed structure {}", name);
        }
    }

    public Structure rotated() {
        if (rotated != null) return rotated;

        BlockInfo[][][] newBlocks = new BlockInfo[blocksZ][blocksY][blocksX];
        for (BlockInfo oldBlock : Arrays.stream(blocks).parallel().flatMap(Arrays::stream).flatMap(Arrays::stream).filter(Objects::nonNull).toList()) {
            Cord cord = oldBlock.cord().rotated();
            newBlocks[(int) cord.x()][(int) cord.y()][(int) cord.z()] = new BlockInfo(cord, oldBlock.material());
        }
        rotated = new Structure(name, newBlocks);
        return rotated;
    }

    public void place(Location location) {
        Utilities.loop(blocks, false, block -> {
            Cord cord = block.cord();
            location.clone().add(cord.x(), cord.y(), cord.z()).getBlock().setType(block.material());
        });
    }

    public boolean isInArena(@NotNull Arena arena) {
        return matchMaterials(arena.currentPlacedMaterials(), materials) && isInRegion(arena.currentPlacedBlocks(), arena.currentPlacedBlocksAmount());
    }

    public boolean matchMaterials(@NotNull Collection<Material> reference, @NotNull Collection<Material> comparison) {
        for (Material material : reference) {
            if (!comparison.contains(material))
                return false;
        }
        return true;
    }

    public boolean isInRegion(BlockInfo @NotNull [][][] region, int regionBlockAmount) {
        if (countElements(blocks) != regionBlockAmount) return false;
        for (int startX = 0; startX <= region.length - blocksX; startX++) {
            for (int startY = 0; startY <= region[0].length - blocksY; startY++) {
                for (int startZ = 0; startZ <= region[0][0].length - blocksZ; startZ++) {
                    if (matchesAtPosition(region, startX, startY, startZ))
                        return true;
                }
            }
        }
        return false;
    }

    private boolean matchesAtPosition(BlockInfo[][][] region, int offsetX, int offsetY, int offsetZ) {
        for (int x = 0; x < blocksX; x++) {
            for (int y = 0; y < blocksY; y++) {
                for (int z = 0; z < blocksZ; z++) {
                    BlockInfo structureBlock = blocks[x][y][z];
                    if (structureBlock != null) {
                        BlockInfo regionBlock = region[offsetX + x][offsetY + y][offsetZ + z];
                        if (regionBlock == null || !regionBlock.material().equals(structureBlock.material()))
                            return false;
                    }
                }
            }
        }
        return true;
    }

    public static int countElements(Object @NotNull [][][] array) {
        int count = 0;
        for (Object[][] x : array) {
            for (Object[] y : x) {
                for (Object z : y) {
                    if (z != null)
                        count++;
                }
            }
        }
        return count;
    }

    public static BlockInfo @NotNull [][][] blocks(World world, @NotNull Cord corner1, @NotNull Cord corner2) {
        Pair<Cord, Cord> corners = Cord.corners(corner1, corner2);
        Cord low = corners.left(), high = corners.right();

        BlockInfo[][][] blocks = new BlockInfo[(int) (high.x() - low.x() + 1)][(int) (high.y() - low.y() + 1)][(int) (high.z() - low.z() + 1)];
        for (int x = (int) low.x(); x <= high.x(); x++) {
            for (int y = (int) low.y(); y <= high.y(); y++) {
                for (int z = (int) low.z(); z <= high.z(); z++) {
                    Block block = world.getBlockAt(x, y, z);
                    if (block.getType().isAir()) continue;
                    blocks[x - (int) low.x()][y - (int) low.y()][z - (int) low.z()] = new BlockInfo(new Cord(x - (int) low.x(), y - (int) low.y(), z - (int) low.z()), block.getType());
                }
            }
        }
        return blocks;
    }
}

File: src/main/kotlin/net/hectus/neobb/matrix/structure/Structure.kt
package net.hectus.neobb.matrix.structure

import com.marcpg.libpg.storing.Cord
import com.marcpg.storage.JsonUtils
import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient
import net.hectus.neobb.NeoBB
import net.hectus.neobb.matrix.BlockInfo
import net.hectus.neobb.matrix.BlockSpace
import net.hectus.neobb.util.Configuration
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.World
import org.bukkit.block.Block
import org.bukkit.inventory.ItemStack
import java.nio.file.Files
import java.nio.file.Path
import kotlin.math.roundToInt

data class PlacedStructure(val structure: Structure, val lastBlock: Block)

@Serializable
open class Structure(
    val name: String,
    val materials: MutableMap<Material, Int> = mutableMapOf(),
    val blocks: BlockSpace = BlockSpace(y = Configuration.MAX_ARENA_HEIGHT),
    @Transient @kotlin.jvm.Transient var rotated: Structure? = null
): java.io.Serializable {
    companion object {
        private fun blocks(world: World, corner1: Cord, corner2: Cord): BlockSpace {
            val corners = Cord.corners(corner1, corner2)
            val low = corners.first
            val high = corners.second

            val blocks = BlockSpace(
                high.x.toInt() - low.x.toInt() + 1,
                high.y.toInt() - low.y.toInt() + 1,
                high.z.toInt() - low.z.toInt() + 1
            )
            for (x in low.x.toInt()..high.x.toInt()) {
                for (y in low.y.toInt()..high.y.toInt()) {
                    for (z in low.z.toInt()..high.z.toInt()) {
                        val block = world.getBlockAt(x, y, z)
                        if (block.type.isAir) continue

                        val cord = Cord(x - low.x, y - low.y, z - low.z)
                        blocks[cord.x.toInt(), cord.y.toInt(), cord.z.toInt()] = BlockInfo(cord, block.type)
                    }
                }
            }
            return blocks
        }
    }

    constructor(name: String, blocks: BlockSpace): this(name, blocks.materials(), blocks, null)
    constructor(name: String, world: World, corner1: Cord, corner2: Cord): this(name, blocks(world, corner1, corner2))

    fun save() {
        if (Configuration.STRUCTURE_MODE == Configuration.StructureMode.SERVER) {
            NeoBB.LOG.warn("Saving structures to the server is not yet supported yet.")
        } else {
            val file: Path = Configuration.STRUCTURE_MODE.pathSupplier.invoke().resolve("$name.json")
            runCatching {
                JsonUtils.save(this, file.toFile())
                NeoBB.LOG.info("Saved structure $name to $file.")
            }.onFailure { e ->
                NeoBB.LOG.error("Failed to save structure $name to $file.", e)
            }
        }
    }

    fun remove() {
        if (Configuration.STRUCTURE_MODE == Configuration.StructureMode.SERVER) {
            NeoBB.LOG.warn("Removing structures from the server is not yet supported yet.")
        } else {
            val file: Path = Configuration.STRUCTURE_MODE.pathSupplier.invoke().resolve("$name.json")
            runCatching {
                Files.delete(file)
                NeoBB.LOG.info("Deleted structure $name from $file.")
            }.onFailure { e ->
                NeoBB.LOG.error("Failed to delete structure $name from $file.", e)
            }
        }
    }

    fun rotated(): Structure {
        if (rotated != null) return rotated!!

        val xLength = blocks.x
        val zLength = blocks.z
        val newBlocks = BlockSpace(zLength, blocks.y, xLength)

        for (y in 0 until blocks.y)
            for (x in 0 until xLength)
                for (z in 0 until zLength)
                    newBlocks[z, y, xLength - 1 - x] = blocks[x, y, z]

        val structure = Structure(name, materials, newBlocks, this)
        rotated = structure
        return structure
    }

    fun place(location: Location, replace: Boolean): Int {
        var total = 0.0
        var obstructed = 0.0
        blocks.forEach { block ->
            total++
            val realBlock: Block = location.clone().add(block!!.cord.x, block.cord.y, block.cord.z).block
            if (realBlock.isEmpty || replace)
                realBlock.type = block.material
            else obstructed++
        }
        return ((obstructed / total) * 100).roundToInt()
    }

    fun isInSpace(space: BlockSpace): Boolean {
        if (materials != space.materials()) return false

        for (startX in 0..space.x - blocks.x) {
            for (startY in 0..space.y - blocks.y) {
                for (startZ in 0..space.z - blocks.z) {
                    if (matchesAtPosition(space, startX, startY, startZ)) return true
                }
            }
        }
        return false
    }

    fun items(): List<ItemStack> = materials.entries.map { ItemStack(it.key, it.value) }

    private fun matchesAtPosition(space: BlockSpace, offsetX: Int, offsetY: Int, offsetZ: Int): Boolean {
        for (x in 0..<blocks.x) {
            for (y in 0..<blocks.y) {
                for (z in 0..<blocks.z) {
                    val structureBlock = blocks[x, y, z] ?: continue

                    val spaceBlock = space[offsetX + x, offsetY + y, offsetZ + z]
                    if (spaceBlock == null || spaceBlock.material != structureBlock.material)
                        return false
                }
            }
        }
        return true
    }
}
